File: swift/Sources/FluentCards/BackgroundImageBuilder.swift
/// BackgroundImageBuilder builds a backgroundImage object for containers and cards.
public final class BackgroundImageBuilder {
    var data: [String: Any] = [:]

    public init() {}

    @discardableResult
    public func withURL(_ url: String) -> BackgroundImageBuilder {
        data["url"] = url
        return self
    }

    @discardableResult
    public func withFillMode(_ fillMode: BackgroundImageFillMode) -> BackgroundImageBuilder {
        data["fillMode"] = fillMode.rawValue
        return self
    }

    @discardableResult
    public func withHorizontalAlignment(_ alignment: HorizontalAlignment) -> BackgroundImageBuilder {
        data["horizontalAlignment"] = alignment.rawValue
        return self
    }

    @discardableResult
    public func withVerticalAlignment(_ alignment: VerticalAlignment) -> BackgroundImageBuilder {
        data["verticalAlignment"] = alignment.rawValue
        return self
    }

    public func build() -> Card {
        return data
    }
}

File: rust/fluentcards/src/builders/background_image_builder.rs
use serde_json::Value;

use crate::enums::*;
use crate::models::Card;

/// Builds a backgroundImage object for containers and cards.
pub struct BackgroundImageBuilder {
    data: Card,
}

impl BackgroundImageBuilder {
    pub(crate) fn new() -> Self {
        BackgroundImageBuilder { data: Card::new() }
    }

    pub fn with_url(&mut self, url: &str) -> &mut Self {
        self.data
            .insert("url".into(), Value::String(url.into()));
        self
    }

    pub fn with_fill_mode(&mut self, fill_mode: BackgroundImageFillMode) -> &mut Self {
        self.data.insert("fillMode".into(), fill_mode.into());
        self
    }

    pub fn with_horizontal_alignment(&mut self, alignment: HorizontalAlignment) -> &mut Self {
        self.data
            .insert("horizontalAlignment".into(), alignment.into());
        self
    }

    pub fn with_vertical_alignment(&mut self, alignment: VerticalAlignment) -> &mut Self {
        self.data
            .insert("verticalAlignment".into(), alignment.into());
        self
    }

    pub fn build(&self) -> Card {
        self.data.clone()
    }
}
